presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java

```diff
@@ -284,36 +283,88 @@ private void checkEncodeInputValidity(List<Integer> input, int maximumBits)
         checkArgument((totalBitLength <= maximumBits), format("The z-address type specified is not large enough to hold %d bits.", totalBitLength));
     }
 
+    /**
+     * The BoundedZValueRange class contains non-null Integer minimum and maximum value lists. The bounds are all inclusive.
+     */
+    private static class BoundedZValueRange
+    {
+        private final List<Integer> minimumValues;
+        private final List<Integer> maximumValues;
+
+        public BoundedZValueRange(List<Integer> minimumValues, List<Integer> maximumValues)
+        {
+            this.minimumValues = minimumValues;
+            this.maximumValues = maximumValues;
+        }
+
+        public Integer getMinimumValueAt(int index)
+        {
+            return minimumValues.get(index);
+        }
+
+        public Integer getMaximumValueAt(int index)
+        {
+            return maximumValues.get(index);
+        }
+    }
+
     private List<ZAddressRange<Long>> zOrderSearchCurve(List<ZValueRange> valueRanges)
     {
-        List<ZAddressRange<Long>> addressRanges = new ArrayList<>();
+        List<BoundedZValueRange> boundedValueRanges = fillUnspecifiedRangesWithDefaults(valueRanges);
 
-        findAddressesRecursively(maxBitLength - 1, 0L, totalBitLength, fillUnspecifiedRangesWithDefaults(valueRanges), addressRanges);
+        int[] lengths = new int[valueRanges.size()];
+        for (int index = 0; index < valueRanges.size(); index++) {
+            lengths[index] = valueRanges.get(index).length;
+        }
```

Comment:
Not sure why we need this. If fillUnspecifiedRangesWithDefaults returns a list of ZValueRange, then we do not need this list, right?

Reply:
Sound you need this later. NVM.

Reply:
Yeah, we can have this array here; but we need a better name for it: IntervalNumbers.

Reply:
We do need the `lengths` list because it is passed into `permuteValueRanges` to run permutations of every interval group. For example, if lengths is [1, 3, 2], we would have a permutation of 
{[0, 0, 0], [0, 0, 1], [0, 1, 0], [0, 1, 1], [0, 2, 0], [0, 2, 1]}.

One thing I could do is return the numOfIntervals from `boundedValueRanges` instead of `valueRanges`, so we operate mainly on only the `BoundedZValueRange` class.